README.md
# Stable Diffusion Dynamic Prompts extension 

A custom extension for [AUTOMATIC1111/stable-diffusion-webui](https://github.com/AUTOMATIC1111/stable-diffusion-webui) that implements an expressive template language for random or combinatorial prompt generation along with features to support deep wildcard directory structures.

<img src="images/extension.png"/>

Using this script, the prompt:

	A {house|apartment|lodge|cottage} in {summer|winter|autumn|spring} by {2$$artist1|artist2|artist3}

Will produce any of the following prompts:

- A **house** in **summer** by **artist1**, **artist2**
- A **lodge** in **autumn** by **artist3**, **artist1**
- A **cottage** in **winter** by **artist2**, **artist3**
- ...

This is especially useful if you are searching for interesting combinations of artists and styles.

You can also pick a random string from a file. Assuming you have the file seasons.txt in WILDCARD_DIR (see below), then:

    __seasons__ is coming

Might generate the following:

- Winter is coming
- Spring is coming
- ...

You can also use the same wildcard twice

    I love __seasons__ better than __seasons__

- I love Winter better than Summer
- I love Spring better than Spring

More complete documentation can be found [here](SYNTAX.md).

## Installation

The extension can be installed directly from within the **Extensions** tab within the Webui
<img src="images/installation.png"/>

You can also install it manually by running the following command from within the webui directory:

	git clone https://github.com/adieyal/sd-dynamic-prompting/ extensions/dynamic-prompts

## Compatible Scripts
Dynamic Prompts works particularly well with the following scripts:
- [randomize](https://github.com/stysmmaker/stable-diffusion-webui-randomize) if you want to randomise non-prompt parameters like steps, width/height, sample, CFG scale, etc. Both scripts can work concurrently.
- X/Y Plot - setting Dynamic Prompts to <a href="#combinatorial-generation">combinatorial mode</a> while using X/Y Plot, lets you exhaustively test prompt and paramter variations simultaneously.


## Template syntax
Documentation can be found [here](SYNTAX.md)

### Fuzzy Glob/recursive wildcard file/directory matching
In addition to standard wildcard tokens such as `__times__` -> `times.txt`, you can also use globbing to match against multiple files at once.
For instance:

`__colors*__` will match any of the following:
- WILDCARD_DIR/colors.txt
- WILDCARD_DIR/colors1.txt
- WILDCARD_DIR/nested/folder/colors1.txt

`__light/**/*__` will match:
- WILDCARD_DIR/nested/folder/light/a.txt
- WILDCARD_DIR/nested/folder/light/b.txt

but won't match
- WILDCARD_DIR/nested/folder/dark/a.txt
- WILDCARD_DIR/a.txt

You can also used character ranges `[0-9]` and `[a-z]` and single wildcard characters `?`. For more examples see [this article](http://pymotw.com/2/glob/).

## Combinatorial Generation
Instead of generating random prompts from a template, combinatorial generation produced every possible prompt from the given string. For example:
`I {love|hate} {New York|Chicago} in {June|July|August}`

will produce:
- I love New York in June
- I love New York in July
- I love New York in August
- I love Chicago in June
- I love Chicago in July
- I love Chicago in August
- I hate New York in June
- I hate New York in July
- I hate New York in August
- I hate Chicago in June
- I hate Chicago in July
- I hate Chicago in August

If a `__wildcard__` is provided, then a new prompt will be produced for every value in the wildcard file. For example:
`My favourite season is __seasons__`

will produce:
- My favourite season is Summer
- My favourite season is August
- My favourite season is Winter
- My favourite season is Sprint

<img src="images/combinatorial_generation.png"/>

You also arbitrarily nest combinations inside wildcards and wildcards in combinations.

Combinatorial generation can be useful if you want to create an image for every artist in a file. It can be enabled by checking the __Combinatorial generation__ checkbox in the ui. Note, __num batches__ changes meaning. With random generation, exactly __num_batches__ * __batch_size__ images are created. With combinatorial generation, at *most* __num_batches__ * __batch_size__ images are created. This upper limit ensures that you don't accidentially create a template that unexpectedly  produces thousands of images.

Combinations are not yet supported, i.e. `{2$$a|b|c}` will treat `2$$a` as one of the options instead of selecting two of a, b and c.

### Combinatorial Batches
The combinatorial batches slider lets you repeat the same set of prompts a number of times with different seeds. The default number of batches is 1.

### Increasing the maximum number of generations
By default, the __Batch count__ silder of  automatic1111 has a maximum value of 100. This can limit the maximum number of generations when using combinatorial generation. You can change the maximum value of this slider by editing ui-config.json and change:

	txt2img/Batch count/maximum": 100

to something larger like:

	txt2img/Batch count/maximum": 1000

## Fixed seed
Select this if you want to use the same seed for every generated image. If there are no wildcards then all the images will be identical. It is useful if you want to test the effect of a particular modifier. For example:

	A beautiful day at the beach __medium/photography/filmtypes__

That way you can isolate the effect of each film type on a particular scene. Here are some of the results:
<img src="images/filmtypes.jpg"/>

## Magic Prompt
Using [Gustavosta](https://huggingface.co/Gustavosta/MagicPrompt-Stable-Diffusion)'s MagicPrompt model, automatically generate new prompts from the input. Trained on 80,000 prompts from [Lexica.art](lexica.art), it can help give you interesting new prompts on a given subject. Here are some automatically generated variations for "dogs playing football":

> dogs playing football, in the streets of a japanese town at night, with people watching in wonder, in the style of studio ghibli and makoto shinkai, highly detailed digital art, trending on artstation

> dogs playing football, in the background is a nuclear explosion. photorealism. hq. hyper. realistic. 4 k. award winning.

> dogs playing football, in the background is a nuclear explosion. photorealistic. realism. 4 k wideshot. cinematic. unreal engine. artgerm. marc simonetti. jc leyendecker

This is compatible with the wildcard syntax described above.

The first time you use it, the model is downloaded. It is approximately 500mb and so will take some time depending on how fast your connection is. It will also take a few seconds on first activation as the model is loaded into memory. Note, if you're low in VRAM, you might get a Cuda error. My GPU uses less than 8GB by YMMV.

<img src="images/magic_prompt.png"/>
You can control the maximum prompt length with the **Max magic prompt length** slider. **Magic prompt creativity** can adjust the generated prompt but you will need to experiment with this setting.

## I'm feeling lucky
Use the [lexica.art](https://lexica.art) API to create random prompts. Useful if you're looking for inspiration, or are simply too lazy to think of your own prompts. When this option is selected, the prompt in the main prompt box is used as a search string. For example, prompt "Mech warrior" might return:

* A large robot stone statue in the middle of a forest by Greg Rutkowski, Sung Choi, Mitchell Mohrhauser, Maciej Kuciara, Johnson Ting, Maxim Verehin, Peter Konig, final fantasy , 8k photorealistic, cinematic lighting, HD, high details, atmospheric,
* a beautiful portrait painting of a ( ( ( cyberpunk ) ) ) armor by simon stalenhag and pascal blanche and alphonse mucha and nekro. in style of digital art. colorful comic, film noirs, symmetry, brush stroke, vibrating colors, hyper detailed. octane render. trending on artstation
* symmetry!! portrait of a robot astronaut, floral! horizon zero dawn machine, intricate, elegant, highly detailed, digital painting, artstation, concept art, smooth, sharp focus, illustration, art by artgerm and greg rutkowski and alphonse mucha, 8 k

<img src="images/feeling-lucky.png">

Leaving the prompt box blank returns a list of completely randomly chosen prompts.

## Attention grabber
This option randomly selects a keyword in your prompt and adds a random amount of emphasis. Below is an example of how this affects the prompt:

	a portrait an anthropomorphic panda mage casting a spell, wearing mage robes, landscape in background, cute, dnd character art portrait, by jason felix and peter mohrbacher, cinematic lighting

<img src="images/emphasis.png">

Tick the __Fixed seed__ checkbox under __Advanced options__ to see how emphasis changes your image without changing seed.


## Write prompts to file
Check the write prompts to file checkbox in order to create a file with all generated prompts. The generated file is a slugified version of the prompt and can be found in the same directory as the generated images, e.g. outputs/txt2img-images
<img src="images/write_prompts.png"/>

## Jinja2 templates
[Jinja2 templates](https://jinja.palletsprojects.com/en/3.1.x/templates/) is an experimental feature that enables you to define prompts imperatively. This is an advanced feature and is only recommended for users who are comfortable writing scripts.

To enable, open the advanced accordion and select __Enable Jinja2 templates__.
<img src="images/jinja_templates.png">

You can read about them in more detail <a href="jinja2.md">here</a>

## WILDCARD_DIR
The extension looks for wildcard files in WILDCARD_DIR. The default location is /path/to/stable-diffusion-webui/extensions/sd-dynamic-prompts/wildcards. It can also be manually defined in the main webui config.json under wildcard_dir. When in doubt, the help text for the extension in the webui lists the full path to WILDCARD_DIR

## Collections
The collections directory contains modifier libraries that you can use as is or to bootstrap your own. Copy the collection that you want to use into the wildcards directory. Note, in previous versions, the collections were stored in the wildcards directory. This has now changed so that your own collections don't get clobbered every time you want to update the extension.
There are currently two collections:

- jumbo
- parrotzone

Jumbo is a very large collection of wildcards across many categories including aesthetics, appearance, artists, medium, style, and time. It is a work in progress, but aims to provide good coverage of various modifier categories. 

Parrotzone is a far smaller and more manageable collection sourced from https://proximacentaurib.notion.site/e28a4f8d97724f14a784a538b8589e7d?v=42948fd8f45c4d47a0edfc4b78937474.

If you're using a Unix/Linux O/S, you can easily create a symlink to the relevant collection rather than copying it across if you don't plan to alter it. E.g.

	ln -sr collections/parrotzone wildcards/


## Contributing
If you're interested in contributing to the development of this extension, here are some features I would like to implement:

1. **Saved templates** - [publicprompts.art](https://publicprompts.art/) produces great prompt templates. e.g. 

> Funky pop Yoda figurine, made of plastic, product studio shot, on a white background, diffused lighting, centered

You can swap out Yoda for anything to create a great character. If I create/discover a great prompt that can be turned into a template, it would be great to have a place to save it. The extension could provide some initial templates. Users could then create and save their own.

2. **Tag-based modifers**. Creating a strict taxonomy of concepts for modifiers is always going to be flawed. Some modifiers fit in multiple categories. It would be better to introduce tags to make it easy to place modifiers in multiple places at once.

3. **Improved modifier management**. The modifier library is a little clunky at the moment. You can find a nested hierarchy of all wildcard files, but it isn't possible to see what's in them or even edit them from within the interface. It might be best to implement this in a separate tab.

4. **Improve the modifier library**. I'm currently creating the library from multiple sources but building taxonomies is hard, especially if you're not a domain expert. The photography section is a good example, there are wildcard files for lighting, filetypes, camera models, perspective, photo websites, etc. Someone with a better understanding of photography might have a better way to divide this. The artists section in particular needs some TLC.

5. **Option tweaking**. There have been a few requests for option tweaking like XY Plot. I'm not sure that that is in scope for this extension, but it would be nice to see if we could find a solution that leverage's XY Plot.

6. **Improved UI**. The current UI works, but a few javascript tweaks could improve the user experience tremendously.

Let me know if there is anything you would be interested in looking into.


CHANGELOG
0.1 Initial implementation
0.2 Fixed Batch img2img implementation
0.3 Added interval implementation for combinations, i.e. you can now use {1-2$$op1|opt2|opt3} which will choose one or two options. Thanks to @galvanized for this.
0.4 Rows in a wildcard file that contain dynamic syntax will be resolved. This enables recusive resolution. See [this](https://github.com/adieyal/sd-dynamic-prompting/issues/3) feature request for a use case
0.4.1 Fixed regular expression for wildcards, it previously didn't allow wildcard files with underscores in them.
0.4.2 Ignoring unknown characters in non-utf8 encoding wildcard files
0.4.3 Printing prompts at logger level INFO
0.4.4 Fixed bug where num_batches was being interpreted as num_images
0.4.5 Enabled generation of grids
0.5.0 The script can now handle nested wildcard directories in order to make it easier to organise large wildcard collections. Comments and empty lines are now ignored. Both features thanks to @evanjs 
0.6.0 Added path globbing for wildcards. Thanks for @evanjs for the idea and original implementation
0.7.0 Updated wildcard ui by @TheAwesomeGoat.7.0 Updated wildcard ui by @TheAwesomeGoat
0.8.0 Combinatorial generation 
0.9.0 Started a modifer library. Turns out it's much better than the long unsorted lists of artists you find elsewhere.
0.9.1 Added local Random instance to ensure repeatability without changing the global random number generator
0.9.2 Added additional artists and refactored code.
0.10.0 Added magic prompts
0.11.0 Migrated to an extension
0.12.0 Added a slider to enable combinatorial generation in batches, in other words, run the same set of prompts with different seeds
0.13.0 Added same seed checkbox to keep the seed constant for all generations.
0.13.1 Fixed light mode css
0.13.2 Fixed combinatorial bug which generated duplicate templates
0.13.3 Added an option to revert 0.9.1. There isn't a ui for it yet, but you can edit constants.py. Change to UNLINK_SEED_FROM_PROMPT = True
0.13.4 Fixed 0.13.3, but thiss time for real
0.13.5 Updated stylesheet from @Yaiol to fix light and dark theme
0.13.6 Moved wildcards into the collections directory and added the parrot zone collection with thanks to @jmpaz 
0.13.7 Python 3.8 compatibility
0.13.8 Allow sampling with replacement for variants, i.e. {2$$X|Y} can return X,X Y,Y X,Y and Y,X. Useful in combination with wildcards. {2$$__artists__} now also chooses two options from the artists.txt file.
0.13.9 Added implicit lower and upper ranges to combinations - thanks to @galvanized
0.14.0 Added the ability to change the combinations joiner from the default "," to anything you like, e.g. {1$$and$$red|blue} gives you red and blue - contributed by @cobryan05 
0.15.0 Improvements to Magic Prompt by @SineSwiper - added a slider to adjust the prompt length and another to tune creativity.
0.15.1 Fixed an error caused when setting wildcard_dir in config.json, also other minor bug fixes
0.15.2 Handle situation where no prompts are generated.
0.16.0 Added a checkbox to enable writing all generated prompts to a text file.
0.17.0 Dynamic prompts is now an AlwaysVisible script. It now plays nicely with other scripts like X/Y plot. Big thanks to @stysmmaker
0.17.1 Minor bug fix checking type of combinatorial batches
0.18.0 Added Jinja2 templates as an experimental feature
0.19.0 Added a checkbox to enable unlinking seeds from prompts. This feature has been available for sometime but there hasn't been a checkbox in the ui to enable it.
0.19.1 Added missing install.py to install Jinja2
0.19.2 Fixed bug in MagicPrompt that broke prompts with pipes, https://github.com/adieyal/sd-dynamic-prompts/issues/54
0.20.0 Added random, randint, and permutations jinja extensions
0.20.1 Using launch.run_pip instead of subprocess to install dependencies
0.21.0 Added a checkbox to disable dynamic prompts if needed
0.22.0 Added a weighted choice function to jinja2 templates
0.23.0 Added I'm feeling lucky for the ultimate in lazy prompting
0.23.1 Fixed a bug that didn't limit the number of prompts when using combinatorial generation and large wildcard files
0.24.0 I'm feeling lucky now using the prompt as a search string.
0.24.1 Re-wrote combinatorial generation to use a queue instead of recursion
0.24.2 Minor bug fix in weighted choice jinja2 extension
0.24.3 Added shuffle to combinatorial generation so that a different subset of prompts is generated every time if the total number of prompts is greater than Batch size * Batch count
0.24.4 Tucked help into separate accordions to prevent too much clutter
0.25.0 Negative prompts are now dynamic, Big thanks to @lilly1987
0.25.1 Explicitly setting encoding when writing prompts to file. Also ignoring errors
0.26.0 Added option to disable the negative prompt
0.27.0 Added option to disable image generation - useful in combination with writing prompts to file
0.27.1 Slight ui re-organisation
0.27.2 Fixed bug when dynamic prompts is disabled - see https://github.com/adieyal/sd-dynamic-prompts/issues/62
0.27.3 Another fix for https://github.com/adieyal/sd-dynamic-prompts/issues/62
0.27.4 Added debug logging
0.28.0 Added tooltips to all components, with thanks to @Eugenii10
0.28.1 Minor magic prompt fixes, with thanks to @SineSwiper
0.29.0 Added attention grabber
0.29.1 Changed attention grabber to use noun phrases
0.29.2 Fixed camera models file in the jumbo collection - see https://github.com/adieyal/sd-dynamic-prompts/issues/69
0.29.3 Fixed localization for help text - see https://github.com/adieyal/sd-dynamic-prompts/issues/68
0.29.4 Fixed pipe (|) in combinations string - see https://github.com/adieyal/sd-dynamic-prompts/issues/54
0.29.5 Added link to documentation on github
0.29.6 Maximum filesize for prompts is 50 to prevent issues on windows
0.29.7 Removed python type aliases for backwards compatibility
0.29.8 Fixed unlink seed from prompt - see https://github.com/adieyal/sd-dynamic-prompts/issues/80

dynamic_prompting.py
from itertools import chain
from pathlib import Path
from typing import Any
from typing import Set
import logging
import math
import os
import pathlib
import re
from random import Random

import gradio as gr

from transformers import AutoTokenizer, AutoModelForCausalLM
from transformers import pipeline
import modules.scripts as scripts
from modules.processing import process_images, fix_seed, Processed
from modules.shared import opts
from modules.devices import get_optimal_device

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

WILDCARD_DIR = getattr(opts, "wildcard_dir", "scripts/wildcards")
MAX_RECURSIONS = 20
VERSION = "0.10.0"
WILDCARD_SUFFIX = "txt"
MAX_IMAGES = 1000

# TODO this needs to be fixed
device = 0 if get_optimal_device() == "cuda" else -1

re_wildcard = re.compile(r"__(.*?)__")
re_combinations = re.compile(r"\{([^{}]*)}")

DEFAULT_NUM_COMBINATIONS = 1

random = Random()

class WildcardFile:
    def __init__(self, path: Path, encoding="utf8"):
        self._path = path
        self._encoding = encoding

    def get_wildcards(self) -> Set[str]:
        is_empty_line = lambda line: line is None or line.strip() == "" or line.strip().startswith("#")

        with self._path.open(encoding=self._encoding, errors="ignore") as f:
            lines = [line.strip() for line in f if not is_empty_line(line)]
            return set(lines)

class WildcardManager:
    def __init__(self, path:str=WILDCARD_DIR):
        self._path = Path(path)

    def _directory_exists(self) -> bool:
        return self._path.exists() and self._path.is_dir()

    def ensure_directory(self) -> bool:
        try:
            self._path.mkdir(parents=True, exist_ok=True)
        except Exception as e:
            logger.exception(f"Failed to create directory {self._path}")

    def get_files(self, relative:bool=False) -> list[Path]:
        if not self._directory_exists():
            return []


        files = self._path.rglob(f"*.{WILDCARD_SUFFIX}")
        if relative:
            files = [f.relative_to(self._path) for f in files]

        return files
    
    def match_files(self, wildcard:str) -> list[WildcardFile]:
        return [
            WildcardFile(path) for path in self._path.rglob(f"{wildcard}.{WILDCARD_SUFFIX}")
        ]
    
    def path_to_wilcard(self, path: Path) -> str:
        rel_path = path.relative_to(self._path)
        return f"__{rel_path.with_suffix('')}__"

    def get_wildcards(self) -> list[str]:
        files = self.get_files(relative=True)
        wildcards = [self.path_to_wilcard(f) for f in files]

        return wildcards

    def get_wildcard_hierarchy(self, path: str):
        path = Path(path)
        files = path.glob("*.txt")
        wildcards = [self.path_to_wilcard(f) for f in files]
        directories = [d for d in path.glob("*") if d.is_dir()]

        hierarchy = {d.name: self.get_wildcard_hierarchy(d) for d in directories}
        return (wildcards, hierarchy)

wildcard_manager = WildcardManager()

class UiCreation:
    def write(self, wildcards: list[str], hierarchy: dict[str, Any]) -> str:
        html = ""
        for wildcard in wildcards:
            html += f"<p>{wildcard}</p>"

        for directory, h in hierarchy.items():
            contents = self.write(h[0], h[1])
            html += f"""
                <button type="button" class="collapsible">{directory} :</button>
                <div class="content">
                    {contents}
                </div>
            """

        return html

    def probe(self) -> str:
        wildcards, hierarchy = wildcard_manager.get_wildcard_hierarchy(WILDCARD_DIR)
        return self.write(wildcards, hierarchy)


ui_creation = UiCreation()

def replace_combinations(match):
    if match is None or len(match.groups()) == 0:
        logger.warning("Unexpected missing combination")
        return ""

class PromptGenerator:
    pass

class MagicPromptGenerator(PromptGenerator):
    generator = None

    def _load_pipeline(self):
        try:
            from modules.safe import unsafe_torch_load, torch
            safe_load = torch.load
            torch.load = unsafe_torch_load

            if MagicPromptGenerator.generator is None:
                tokenizer = AutoTokenizer.from_pretrained("Gustavosta/MagicPrompt-Stable-Diffusion")
                model = AutoModelForCausalLM.from_pretrained("Gustavosta/MagicPrompt-Stable-Diffusion")

                MagicPromptGenerator.tokenizer = tokenizer
                MagicPromptGenerator.model = model

                MagicPromptGenerator.generator = pipeline(task="text-generation", model=model, tokenizer=tokenizer, device=device)

            return MagicPromptGenerator.generator
        finally:
            torch.load = safe_load

        return MagicPromptGenerator.generator

    def __init__(self, prompt_generator: PromptGenerator):
        self._generator = self._load_pipeline()
        self._prompt_generator = prompt_generator

    def generate(self, *args, **kwargs) -> str:
        prompts = self._prompt_generator.generate(*args, **kwargs)
        new_prompts = [self._generator(prompt)[0]["generated_text"] for prompt in prompts]

        for old_prompt, magic_prompt in zip(prompts, new_prompts):
            print(f"Generated magic prompt for {old_prompt}: {magic_prompt}")

        return new_prompts

class RandomPromptGenerator(PromptGenerator):
    def __init__(self, template):
        self._template = template

    def _replace_combinations(self, match):
        if match is None or len(match.groups()) == 0:
            logger.warning("Unexpected missing combination")
            return ""

        variants = [s.strip() for s in match.groups()[0].split("|")]
        if len(variants) > 0:
            first = variants[0].split("$$")
            quantity = DEFAULT_NUM_COMBINATIONS
            if len(first) == 2: # there is a $$
                prefix_num, first_variant = first
                variants[0] = first_variant
                
                try:
                    prefix_ints = [int(i) for i in prefix_num.split("-")]
                    if len(prefix_ints) == 1:
                        quantity = prefix_ints[0]
                    elif len(prefix_ints) == 2:
                        prefix_low = min(prefix_ints)
                        prefix_high = max(prefix_ints)
                        quantity = random.randint(prefix_low, prefix_high)
                    else:
                        raise 
                except Exception:
                    logger.warning(f"Unexpected combination formatting, expected $$ prefix to be a number or interval. Defaulting to {DEFAULT_NUM_COMBINATIONS}")
            
            try:
                picked = random.sample(variants, quantity)
                return ", ".join(picked)
            except ValueError as e:
                logger.exception(e)
                return ""

        return ""

    def _replace_wildcard(self, match):
        if match is None or len(match.groups()) == 0:
            logger.warning("Expected match to contain a filename")
            return ""

        wildcard = match.groups()[0]
        wildcard_files = wildcard_manager.match_files(wildcard)

        if len(wildcard_files) == 0:
            logging.warning(f"Could not find any wildcard files matching {wildcard}")
            return ""

        wildcards = set().union(*[f.get_wildcards() for f in wildcard_files])

        if len(wildcards) > 0:
            return random.choice(list(wildcards))
        else:
            logging.warning(f"Could not find any wildcards in {wildcard}")
            return ""
    

    def pick_variant(self, template):
        if template is None:
            return None

        return re_combinations.sub(lambda x: self._replace_combinations(x), template)

    def pick_wildcards(self, template):
        return re_wildcard.sub(lambda x: self._replace_wildcard(x), template)

    def generate_prompt(self, template):
        old_prompt = template
        counter = 0
        while True:
            counter += 1
            if counter > MAX_RECURSIONS:
                raise Exception("Too many recursions, something went wrong with generating the prompt")

            prompt = self.pick_variant(old_prompt)
            prompt = self.pick_wildcards(prompt)

            if prompt == old_prompt:
                logger.info(f"Prompt: {prompt}")
                return prompt
            old_prompt = prompt

    def generate(self, num_prompts) -> list[str]:
        all_prompts = [
            self.generate_prompt(self._template) for _ in range(num_prompts)
        ]

        return all_prompts

class CombinatorialPromptGenerator(PromptGenerator):
    def __init__(self, template):
        self._template = template

    def generate_from_variants(self, seed_template):
        templates = [seed_template]
        new_templates = []
        variants = re_combinations.findall(templates[0])
        for variant in variants:
            for val in variant.split("|"):
                for template in templates:
                    new_templates.append(template.replace(f"{{{variant}}}", val, 1))
            templates = new_templates
            new_templates = []

        if len(templates) == 0:
            return [seed_template]
        return templates

    def generate_from_wildcards(self, seed_template, recursion=0):
        templates = []

        if recursion > MAX_RECURSIONS:
            raise Exception("Too many recursions, something went wrong with generating the prompt: " + seed_template)

        template = seed_template
        wildcards = re_wildcard.findall(template)
        if len(wildcards) == 0:
            return [template]

        for wildcard in wildcards:
            wildcard_files = wildcard_manager.match_files(wildcard)
            for val in chain(*[f.get_wildcards() for f in wildcard_files]):
                new_template = template.replace(f"__{wildcard}__", val, 1)
                logging.debug(f"New template: {new_template}")
                templates.append(new_template)

        new_templates = []
        for template in templates:
            new_templates += self.generate_from_wildcards(template, recursion=recursion + 1)

        return new_templates


    def generate(self, max_prompts=MAX_IMAGES) -> list[str]:
        templates = [self._template]
        all_prompts = []

        while True:
            if len(templates) == 0 or len(all_prompts) > max_prompts:
                break

            template = templates.pop(0)
            new_prompts = self.generate_from_wildcards(template)
            templates.extend(new_prompts)

            template = templates.pop(0)
            new_prompts = self.generate_from_variants(template)
            no_new_prompts = len(new_prompts) == 1

            if no_new_prompts:
                all_prompts.append(new_prompts[0])
            else:
                templates.extend(new_prompts)

        return all_prompts[:max_prompts]

class Script(scripts.Script):
    def title(self):
        return f"Dynamic Prompting v{VERSION}"

    def ui(self, is_img2img):
        wildcard_html = ui_creation.probe()
        html = f"""
            <style>
                .collapsible {{
                    background-color: #1f2937;
                    color: white;
                    cursor: pointer;
                    padding: 18px;
                    width: 100%;
                    border: 2px #0C111C;
                    border-right-style: solid;
                    border-top-style: solid;
                    border-left-style: solid;
                    border-bottom-style: solid;
                    border-radius: 8px 8px 8px 8px;
                    padding: 5px;
                    margin-top: 10px;
                    text-align: left;
                    outline: none;
                    font-size: 15px;
                }}

                .active, .collapsible:hover {{
                    background-color: #555;
                }}

                .codeblock {{
                    background-color: #06080D;
                }}

                .content {{
                    padding: 0 18px;
                    display: none;
                    overflow: hidden;
                    border: 2px #0C111C;
                    border-right-style: solid;
                    border-bottom-style: solid;
                    border-left-style: solid;
                    border-radius: 0px 0px 8px 8px;
                    background-color: #1f2937;
                }}

                #is-combinatorial:after {{
                    content: "Generate all possible prompts up to a maximum of Batch count * Batch size)"
                }}

                #is-magicprompt:after {{
                    content: "Automatically update your prompt with interesting modifiers. (Runs slowly the first time)"
                }}
            </style>


            <h3><strong>Combinations</strong></h3>
            Choose a number of terms from a list, in this case we choose two artists: 
            <code class="codeblock">{{2$$artist1|artist2|artist3}}</code>

            If $$ is not provided, then 1$$ is assumed.

            A range can be provided:
            <code class="codeblock">{{1-3$$artist1|artist2|artist3}}</code>
            In this case, a random number of artists between 1 and 3 is chosen.

            <br/><br/>

            <h3><strong>Wildcards</strong></h3>
            {wildcard_html}

			<br/>
            If the groups wont drop down click <strong onclick="check_collapsibles()" style="cursor: pointer">here</strong> to fix the issue.

            <br/><br/>

            <code class="codeblock">WILDCARD_DIR: {WILDCARD_DIR}</code><br/>
            <small onload="check_collapsibles()">You can add more wildcards by creating a text file with one term per line and name is mywildcards.txt. Place it in {WILDCARD_DIR}. <code class="codeblock">__&#60;folder&#62;/mywildcards__</code> will then become available.</small>
        """
        is_combinatorial = gr.Checkbox(label="Combinatorial generation", value=False, elem_id="is-combinatorial")
        is_magic_prompt = gr.Checkbox(label="Magic prompt", value=False, elem_id="is-magicprompt")
        info = gr.HTML(html)
        return [info, is_combinatorial, is_magic_prompt]

    def run(self, p, info, is_combinatorial, is_magic_prompt):
        fix_seed(p)

        original_prompt = p.prompt[0] if type(p.prompt) == list else p.prompt
        original_seed = p.seed
        random.seed(original_seed)

        if is_combinatorial:
            prompt_generator = CombinatorialPromptGenerator(original_prompt)
        else:
            prompt_generator = RandomPromptGenerator(original_prompt)

        if is_magic_prompt:
            prompt_generator = MagicPromptGenerator(prompt_generator)
        
        num_images = p.n_iter * p.batch_size
        all_prompts = prompt_generator.generate(num_images)
        updated_count = len(all_prompts)
        p.n_iter = math.ceil(updated_count / p.batch_size)

        all_seeds = [int(p.seed) + (x if p.subseed_strength == 0 else 0) for x in range(updated_count)]

        logger.info(f"Prompt matrix will create {updated_count} images in a total of {p.n_iter} batches.")

        p.prompt = all_prompts
        p.seed = all_seeds

        p.prompt_for_display = original_prompt
        processed = process_images(p)

        p.prompt = original_prompt
        p.seed = original_seed

        return processed

wildcard_manager.ensure_directory()
